# TV 'Golden Age' Analysis <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, I'll be delving into the fascinating world of the entertainment industry, particularly focusing on television during what many consider to be its "Golden Age." It all began in 1999 with the groundbreaking release of *The Sopranos*, and the wave of exceptional television content hasn't stopped since.

My objective is to examine a dataset packed with information on various movies and TV shows. Specifically, I'm interested in understanding how the number of votes a title garners correlates with its ratings. The hypothesis driving this investigation is that during this illustrious "Golden Age," highly-rated TV shows tend to attract the most attention and therefore accumulate the most votes.

### Stages 
I have access to a dataset stored in the `/datasets/movies_and_shows.csv` file, containing information about both movies and shows. Initially, there's no indication of the data's quality, so my first step is to thoroughly examine it.

My project unfolds in three distinct stages:
1) Initial Data Assessment: Here, I'll evaluate the quality and relevance of the data, identifying any potential issues or inconsistencies.
2) Data Cleansing: In this phase, I'll address any significant problems uncovered during the assessment. This involves preprocessing the data to ensure its accuracy and reliability.
3) Analytical Exploration: Once the data is refined, I'll dive into the analysis phase. Here, I'll extract insights and draw conclusions from the dataset, exploring various relationships and patterns within the data.

## Stage 1. Data overview <a id='data_review'></a>

In [719]:
import pandas as pd

In [720]:
df = pd.read_csv('/datasets/movies_and_shows.csv')

In [721]:
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

In [722]:
# Obtaining general information about the data in df
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


release Year    imdb sc0re    imdb v0tes
count  85579.000000  80970.000000  8.085300e+04
mean    2015.879994      6.425877  5.978271e+04
std        7.724668      1.122655  1.846287e+05
min     1954.000000      1.500000  5.000000e+00
25%     2015.000000      5.700000  1.266000e+03
50%     2018.000000      6.500000  5.448000e+03
75%     2021.000000      7.200000  3.360900e+04
max     2022.000000      9.500000  2.294231e+06

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>

In [723]:
# The list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

In [724]:
# renaming columns
df = df.rename(
    columns={
        '   name': 'name',
        'Character': 'character',
        'r0le': 'role',
        'TITLE': 'title',
        '  Type': 'type',
        'release Year': 'release_year',
        'imdb sc0re': 'imdb_score',
        'imdb v0tes': 'imdb_votes'
    }
)

In [725]:
# Checking result: The list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>

In [726]:
# Calculating missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [727]:
# Dropping rows where columns with title, scores and votes have missing values
df = df.dropna()

In [728]:
# Counting missing values
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>

In [729]:
# Counting duplicate rows
df.duplicated().sum()

6994

In [730]:
# Produce table with duplicates (with original rows included) and review last 5 rows
duplicated_rows = df[df.duplicated()]
duplicated_rows

name                  character      role  \
7561         Philip Greene  Baseball Fan (uncredited)     ACTOR   
14512             Dan Levy                   Reporter     ACTOR   
18952     Nicolas Le Nev??                    unknown  DIRECTOR   
22456    John Iii Franklin                    Himself     ACTOR   
29557         Claudio Roca                   Nicol?­s     ACTOR   
...                    ...                        ...       ...   
85569       Jessica Cediel            Liliana Navarro     ACTOR   
85570  Javier Gardeaz?­bal   Agust??n "Peluca" Ort??z     ACTOR   
85571        Carla Giraldo             Valery Reinoso     ACTOR   
85572  Ana Mar??a S?­nchez                    Lourdes     ACTOR   
85577         Isabel Gaona                     Cacica     ACTOR   

                                 title   type  release_year  \
7561                   How Do You Know  MOVIE          2010   
14512  A Very Harold & Kumar Christmas  MOVIE          2011   
18952                       Sammy & Co   SHOW          2014   
22456                    Last Chance U   SHOW          2016   
29557                   Narcos: Mexico   SHOW          2018   
...                                ...    ...           ...   
85569                          Lokillo  MOVIE          2021   
85570                          Lokillo  MOVIE          2021   
85571                          Lokillo  MOVIE          2021   
85572                          Lokillo  MOVIE          2021   
85577                          Lokillo  MOVIE          2021   

                                 genres  imdb_score  imdb_votes  
7561     ['comedy', 'drama', 'romance']         5.4     50383.0  
14512  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
18952         ['animation', 'european']         5.7        31.0  
22456        ['documentation', 'sport']         8.4      6897.0  
29557                ['drama', 'crime']         8.4     82042.0  
...                                 ...         ...         ...  
85569                        ['comedy']         3.8        68.0  
85570                        ['comedy']         3.8        68.0  
85571                        ['comedy']         3.8        68.0  
85572                        ['comedy']         3.8        68.0  
85577                        ['comedy']         3.8        68.0  

[6994 rows x 9 columns]

In [731]:
# Removing duplicate rows
df = df.drop_duplicates()

In [732]:
# Checking for duplicates
df.duplicated().sum()

0

In [733]:
# Viewing unique type names
implicit_duplicates = df['type'].unique()
implicit_duplicates

array(['MOVIE', 'the movie', 'tv show', 'shows', 'movies', 'tv shows',
       'tv series', 'tv', 'SHOW'], dtype=object)

In [734]:
# Function for replacing implicit duplicates
def replace_wrong_show(wrong_shows_list, correct_show):
    df['type'] = df['type'].replace(wrong_shows_list, correct_show)
    return df['type']

In [735]:
# Removing implicit duplicates
replace_wrong_show(implicit_duplicates, 'SHOW')

0        SHOW
1        SHOW
2        SHOW
3        SHOW
4        SHOW
         ... 
85573    SHOW
85574    SHOW
85575    SHOW
85576    SHOW
85578    SHOW
Name: type, Length: 73859, dtype: object

In [736]:
# Viewing unique genre names
df['type'].unique()

array(['SHOW'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 
The absence of duplicates will make the results more precise and easier to understand.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous stages of the project, I'm now ready to outline how I'll test the assumption. I'll start by calculating the average number of votes for each score, using data from the `imdb_score` and `imdb_votes` columns. Then, I'll compare these averages to see how they relate to each other. If the averages for shows with the highest scores are greater than those for shows with lower scores, it suggests that the assumption holds true.

Here's how I'll proceed:
1) Filtering the Dataframe: I'll narrow down the dataset to include only shows released in 1999 or later.
2) Bucketing Scores: To simplify our analysis and make the results clearer, I'll group scores into buckets by rounding the values to the nearest integer. This approach will maintain the integrity of our research while enhancing interpretability.
3) Identifying and Handling Outliers: I'll identify any outliers among the scores based on the number of votes they've received. Scores with insufficient votes may skew our results, so I'll exclude them from our analysis.
4) Calculating Average Votes: With the data appropriately filtered and outliers removed, I'll calculate the average number of votes for each score bucket. This will allow me to assess whether the assumption aligns with the observed results.

In [737]:
# Using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
millenium = df[(df["release_year"] >= 1999) & (df["type"] == "SHOW")]

In [740]:
# Use groupby() for scores and count all unique values in each group, print the result
millenium.groupby('imdb_score').count()

name  character   role  title   type  release_year  genres  \
imdb_score                                                                
2.0           207        207    207    207    207           207     207   
3.0           562        562    562    562    562           562     562   
4.0          3169       3169   3169   3169   3169          3169    3169   
5.0          8926       8926   8926   8926   8926          8926    8926   
6.0         23201      23201  23201  23201  23201         23201   23201   
7.0         21517      21517  21517  21517  21517         21517   21517   
8.0         11568      11568  11568  11568  11568         11568   11568   
9.0           723        723    723    723    723           723     723   
10.0            8          8      8      8      8             8       8   

            imdb_votes  
imdb_score              
2.0                207  
3.0                562  
4.0               3169  
5.0               8926  
6.0              23201  
7.0              21517  
8.0              11568  
9.0                723  
10.0                 8

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

In [741]:
# Filter dataframe using two conditions (scores to be in the range 4-9)
millenium = millenium[millenium['imdb_score'] >= 4]
millenium = millenium[millenium['imdb_score'] <= 9]

# Group scores and corresponding average number of votes, reset index and print the result
result = millenium.groupby('imdb_score').mean().reset_index()
result

imdb_score  release_year     imdb_votes
0         4.0   2017.836541    8699.529820
1         5.0   2017.242326   14236.204123
2         6.0   2017.154735   28807.336192
3         7.0   2016.370823   51308.138542
4         8.0   2015.521352  145532.023081
5         9.0   2014.863071  251330.439834

In [742]:
# Round column with averages
result = result.round()

# Rename columns
result = result.rename(columns={
    'imdb_votes': 'average_imdb_votes'
})

# Print dataframe in descending order
result.sort_values(by='average_imdb_votes', ascending=False)

imdb_score  release_year  average_imdb_votes
5         9.0        2015.0            251330.0
4         8.0        2016.0            145532.0
3         7.0        2016.0             51308.0
2         6.0        2017.0             28807.0
1         5.0        2017.0             14236.0
0         4.0        2018.0              8700.0

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The conducted research validates that during the "Golden Age" of television, highly-rated shows tend to accumulate the highest number of votes. Notably, while shows with a score of 4 garner more votes than those with scores 5 and 6, the top three score categories (7-9) boast the most substantial number of votes. It's worth noting that the analyzed data covers approximately 94% of the original dataset, instilling confidence in the reliability of our findings.